url72c32f3c54_Quickhull3d_quickhull3d_tgz-pJ8-com_github_quickhull3d_SimpleExampleJ8/src/com/github/quickhull3d/Face.java

```diff
@@ -0,0 +1,533 @@
+    public double areaSquared(HalfEdge hedge0, HalfEdge hedge1) {
+        Point3d p0 = hedge0.tail().pnt;
+        Point3d p1 = hedge0.head().pnt;
+        Point3d p2 = hedge1.head().pnt;
+
+        double dx1 = p1.x - p0.x;
+        double dy1 = p1.y - p0.y;
+        double dz1 = p1.z - p0.z;
+
+        double dx2 = p2.x - p0.x;
+        double dy2 = p2.y - p0.y;
+        double dz2 = p2.z - p0.z;
+
+        double x = dy1 * dz2 - dz1 * dy2;
+        double y = dz1 * dx2 - dx1 * dz2;
+        double z = dx1 * dy2 - dy1 * dx2;
+
+        return x * x + y * y + z * z;
+    }
+
+    private void computeNormalAndCentroid() {
+        computeNormal(normal);
+        computeCentroid(centroid);
+        planeOffset = normal.dot(centroid);
+        int numv = 0;
+        HalfEdge he = he0;
+        do {
+            numv++;
+            he = he.next;
+        } while (he != he0);
+        if (numv != numVerts) {
+            throw new InternalErrorException("face " + getVertexString() + " numVerts=" + numVerts + " should be " + numv);
+        }
+    }
+
+    private void computeNormalAndCentroid(double minArea) {
+        computeNormal(normal, minArea);
+        computeCentroid(centroid);
+        planeOffset = normal.dot(centroid);
+    }
+
+    private Face connectHalfEdges(HalfEdge hedgePrev, HalfEdge hedge) {
+        Face discardedFace = null;
+
+        if (hedgePrev.oppositeFace() == hedge.oppositeFace()) { // then there is
+                                                                // a redundant
+                                                                // edge that we
+                                                                // can get rid
+                                                                // off
+
+            Face oppFace = hedge.oppositeFace();
+            HalfEdge hedgeOpp;
+
+            if (hedgePrev == he0) {
+                he0 = hedge;
+            }
+            if (oppFace.numVertices() == 3) { // then we can get rid of the
+                                              // opposite face altogether
+                hedgeOpp = hedge.getOpposite().prev.getOpposite();
+
+                oppFace.mark = DELETED;
+                discardedFace = oppFace;
+            } else {
+                hedgeOpp = hedge.getOpposite().next;
+
+                if (oppFace.he0 == hedgeOpp.prev) {
+                    oppFace.he0 = hedgeOpp;
+                }
+                hedgeOpp.prev = hedgeOpp.prev.prev;
+                hedgeOpp.prev.next = hedgeOpp;
+            }
+            hedge.prev = hedgePrev.prev;
+            hedge.prev.next = hedge;
+
+            hedge.opposite = hedgeOpp;
+            hedgeOpp.opposite = hedge;
+
+            // oppFace was modified, so need to recompute
+            oppFace.computeNormalAndCentroid();
+        } else {
+            hedgePrev.next = hedge;
+            hedge.prev = hedgePrev;
+        }
+        return discardedFace;
+    }
+
+    void checkConsistency() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.